keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,597 @@
+```
+This way the cluster would work consistently because there is a single ClusterVersion across the whole cluster.
+
+## Design Details
+
+On high level, a cluster feature gate would need:
+1. a way to [register new feature gates](#register-new-feature-gates) when a developer adds new features.
+1. a way to [set the feature gates](#set-the-feature-gates) when each member starts up.
+1. a [consensus algorithm](#consensus-algorithm) to determine if a feature is enabled for the whole cluster, even in mixed version scenarios.
+1. an [interface](#interface-change-for-server-code) to query if a feature is enabled for the whole cluster to be used in the server code.
+1. [client APIs](#client-apis-changes) to query if a feature is enabled for the whole cluster.
+1. a way to [remove a feature gate][#feature-removal] when it is no longer useful or have graduated.
+
+### Register New Feature Gates
+
+A feature can be registered as server level feature or cluster level feature, but not both.
+For data compatibility, we do not support changing a feature gate from one type to another. 
+If a developer finds out that their feature needs to be a cluster level feature instead of server level after a release, they should add a new feature and use a new feature name.
+
+NOTE: even though we have `PreRelease` information in the map, the purpose of cluster feature gate is not for lifecycle management. It is mainly for cluster level feature enablement. So a cluster feature can be `false` even after `GA`, and may not be removed if we want to keep its togglability. 
+This is different from server level feature gate, where the feature gate is used for lifecycle tracking purpose, and a feature gate is expected to be removed eventually and permanent boolean flags would be added if needed. 
+
+```go
+DefaultEtcdServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+DefaultEtcdClusterFeatureGates := map[Feature]VersionedSpecs {
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+### Set the Feature Gates
+To set the correct user expectation, we will add a new flag `--cluster-feature-gates` (+`config-file` configuration support) similar to the server `--feature-gates` to set the values of cluster feature gates as expected by the local server. The server will use this flag value to set the `proposed_cluster_params` sent to the cluster. The final values of the cluster feature gate are actually decided through raft consensus.
```

Comment:
I would not distinguish cluster feature flags in command line. It's still a feature, that is gated, the cluster part only changes it's enablement moment.